src/plan.rs

```diff
@@ -0,0 +1,1251 @@
+                (stack, self.descriptor.unsigned_script_sig())
+            }
+        })
+    }
+
+    /// Update a PSBT input with the metadata required to complete this plan
+    ///
+    /// This will only add the metadata for items required to complete this plan. For example, if
+    /// there are multiple keys present in the descriptor, only the few used by this plan will be
+    /// added to the PSBT.
+    pub fn update_psbt_input(&self, input: &mut psbt::Input) {
+        if let Descriptor::Tr(tr) = &self.descriptor {
+            #[derive(Default)]
+            struct TrDescriptorData {
+                tap_script: Option<Script>,
+                control_block: Option<ControlBlock>,
+                internal_key: Option<XOnlyPublicKey>,
+                key_origins: BTreeMap<XOnlyPublicKey, (Vec<TapLeafHash>, bip32::KeySource)>,
+            }
+
+            let spend_info = tr.spend_info();
+            input.tap_merkle_root = spend_info.merkle_root();
+
+            let data = self
+                .template
+                .iter()
+                .fold(TrDescriptorData::default(), |mut data, item| {
+                    match item {
+                        Placeholder::TapScript(script) => data.tap_script = Some(script.clone()),
+                        Placeholder::TapControlBlock(cb) => data.control_block = Some(cb.clone()),
+                        Placeholder::SchnorrSig(pk, sig_type, _) => {
+                            let raw_pk = pk.to_x_only_pubkey();
+
+                            let leaf_hash = match sig_type {
+                                SchnorrSigType::KeySpend { .. } => {
+                                    data.internal_key = Some(raw_pk);
+                                    None
+                                }
+                                SchnorrSigType::ScriptSpend { leaf_hash } => Some(leaf_hash),
+                            };
+
+                            data.key_origins
+                                .entry(raw_pk)
+                                .and_modify(|(tapleaf_hashes, _)| {
+                                    if let Some(leaf_hash) = leaf_hash {
+                                        tapleaf_hashes.push(*leaf_hash);
+                                    }
+                                })
+                                .or_insert_with(|| {
+                                    (
+                                        if let Some(lh) = leaf_hash {
+                                            vec![*lh]
+                                        } else {
+                                            vec![]
+                                        },
+                                        (pk.master_fingerprint(), pk.full_derivation_path()),
+                                    )
+                                });
+                        }
+                        _ => {}
+                    }
+
+                    data
+                });
```

Comment:
I can't figure out a way to enforce with the type system that all the placeholders refer to the same leaf hash. I think it's guaranteed by the current implementation, but not enforced by the types.

I will update this code that updates the psbt so that it assumes all the placeholders are aligned and panics otherwise.

Reply:
Addressed in 503b3f8